client/consul_template.go

```diff
@@ -191,28 +278,85 @@ WAIT:
 			}
 
 			break WAIT
-		}
-	}
+		case <-tm.runner.RenderEventCh():
+			events := tm.runner.RenderEvents()
+			joinedSet := make(map[string]struct{})
+			for _, event := range events {
+				missing := event.MissingDeps
+				if missing == nil {
+					continue
+				}
 
-	// Read environment variables from env templates
-	envMap, err := loadTemplateEnv(tm.templates, taskDir)
-	if err != nil {
-		tm.hook.Kill("consul-template", err.Error(), true)
-		return
-	}
-	envBuilder.SetTemplateEnv(envMap)
+				for _, dep := range missing.List() {
+					joinedSet[dep.String()] = struct{}{}
+				}
+			}
 
-	allRenderedTime = time.Now()
-	tm.hook.UnblockStart("consul-template")
+			// Check to see if the new joined set is the same as the old
+			different := len(joinedSet) != len(missingDependencies)
+			if !different {
+				for k := range joinedSet {
+					if _, ok := missingDependencies[k]; !ok {
+						different = true
+						break
+					}
+				}
+			}
 
-	// If all our templates are change mode no-op, then we can exit here
-	if tm.allTemplatesNoop() {
-		return
+			// Nothing to do
+			if !different {
+				continue
+			}
+
+			// Update the missing set
+			missingDependencies = joinedSet
+
+			// Update the event timer channel
+			if eventTimer == nil {
+				// We are creating the event channel for the first time.
+				eventTimer = time.NewTimer(tm.config.MaxTemplateEventRate)
+				eventTimerCh = eventTimer.C
+				defer eventTimer.Stop()
+				outstandingEvent = true
+				continue
```

Comment:
If you create a stopped timer outside of this loop could you kill this block and just let the `if !outstandEvent {` block handle resetting the timer?